internal/restore/pvc_action.go

```diff
@@ -228,3 +292,145 @@ func getDataUploadResult(ctx context.Context, restore *velerov1api.Restore, pvc
+		Spec: velerov2alpha1.DataDownloadSpec{
+			TargetVolume: velerov2alpha1.TargetVolumeSpec{
+				PVC:       pvc.Name,
+				Namespace: pvc.Namespace,
+			},
+			BackupStorageLocation: dataUploadResult.BackupStorageLocation,
+			DataMover:             dataUploadResult.DataMover,
+			SnapshotID:            dataUploadResult.SnapshotID,
+			SourceNamespace:       dataUploadResult.SourceNamespace,
+			OperationTimeout:      restore.Spec.ItemOperationTimeout,
+		},
+	}
+
+	return dataDownload
+}
+
+func restoreFromVolumeSnapshot(pvc *corev1api.PersistentVolumeClaim, snapClient snapshotterClientSet.Interface,
+	logger logrus.FieldLogger) error {
+	volumeSnapshotName, ok := pvc.Annotations[util.VolumeSnapshotLabel]
+	if !ok {
+		logger.Info("Skipping PVCRestoreItemAction for PVC , PVC does not have a CSI volumesnapshot.")
+		return nil
+	}
+
+	vs, err := snapClient.SnapshotV1().VolumeSnapshots(pvc.Namespace).Get(context.TODO(), volumeSnapshotName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, fmt.Sprintf("Failed to get Volumesnapshot %s/%s to restore PVC %s/%s", pvc.Namespace, volumeSnapshotName, pvc.Namespace, pvc.Name))
+	}
+
+	if _, exists := vs.Annotations[util.VolumeSnapshotRestoreSize]; exists {
+		restoreSize, err := resource.ParseQuantity(vs.Annotations[util.VolumeSnapshotRestoreSize])
+		if err != nil {
+			return errors.Wrapf(err, fmt.Sprintf("Failed to parse %s from annotation on Volumesnapshot %s/%s into restore size",
+				vs.Annotations[util.VolumeSnapshotRestoreSize], vs.Namespace, vs.Name))
+		}
+		// It is possible that the volume provider allocated a larger capacity volume than what was requested in the backed up PVC.
+		// In this scenario the volumesnapshot of the PVC will end being larger than its requested storage size.
+		// Such a PVC, on restore as-is, will be stuck attempting to use a Volumesnapshot as a data source for a PVC that
+		// is not large enough.
+		// To counter that, here we set the storage request on the PVC to the larger of the PVC's storage request and the size of the
+		// VolumeSnapshot
+		setPVCStorageResourceRequest(pvc, restoreSize, logger)
+	}
+
+	resetPVCSpec(pvc, volumeSnapshotName)
+
+	return nil
+}
+
+func restoreFromDataUploadResult(ctx context.Context, restore *velerov1api.Restore, pvc *corev1api.PersistentVolumeClaim,
+	operationID string, kubeClient kubernetes.Interface, veleroClient veleroClientSet.Interface) (*velerov2alpha1.DataDownload, error) {
+	dataUploadResult, err := getDataUploadResult(ctx, restore, pvc, kubeClient)
+	if err != nil {
+		return nil, errors.Wrapf(err, "fail get DataUploadResult for restore: %s", restore.Name)
+	}
+	pvc.Spec.VolumeName = ""
+	if pvc.Spec.Selector == nil {
+		pvc.Spec.Selector = &metav1.LabelSelector{}
+	}
+	if pvc.Spec.Selector.MatchLabels == nil {
+		pvc.Spec.Selector.MatchLabels = make(map[string]string)
+	}
+	pvc.Spec.Selector.MatchLabels[util.DynamicPVRestoreLabel] = label.GetValidName(fmt.Sprintf("%s.%s", pvc.Namespace, pvc.Name))
```

Comment:
We may also need to enhance here, think below scenarios:
One restore is successfully created from the backup, so there is a PV with label `<pvc.Namespace>.<pvc.Name>`
Users delete the restored namespace
For some reason, the PV is not successfully deleted
Now do a new restore to the same namesapce, a new PVC is created to be bound by matching the same label `<pvc.Namespace>.<pvc.Name>`
Since there is already a PV with label `<pvc.Namespace>.<pvc.Name>`, the PVC is bound immediately

Reply:
Then we may also need to add the Restore UID here.

Reply:
Will that be too long for a PV name? Can we remove the pvc.Namespace since for a particular restore, the target namespace should always be the same?

Reply:
That should work, but we need to take care of two labels all the time. The label `DynamicPVRestoreLabel` is well known since it has been mentioned in the design, so we will take care of it during other backup types, e.g., during PVB, the label will be removed. However, `RestoreUIDLabel` is not mentioned in the design and it is not a good practice to have many labels for the same purpose from the perspective of the design.

Therefore, do we have any other way to make it?

Reply:
After discussion, still using the label `DynamicPVRestoreLabel`. The format is changed to `<pvc-namespace>.<pvc-name>.<random-string(length is 5)>`